### PR TITLE
Persist state

### LINKED ISSUE #30: Persist backend state across server restarts

### PR DESCRIPTION
**Pull request by richardolsson**
This PR intends to resolve #30.

This contained a known bug:

When there are multiple projects, the state from store.json doesn't get loaded into all projects if the user only interacts with one of them. This means that data can get lost, if a user _updates_ one project without first _visiting_ all projects.

6ac5ceb tries to fix the bug

### CODE REVIEW

File: webapp/src/store/Store.ts
Comment (WULCAN) [inline webapp/src/store/Store.ts:21]: While the first call to `getProjectStore` is awaiting `loadFromDisk`, concurrent calls will find `globalThis.store` true-like and continue working with an empty object as `initialState`.

These concurrent calls will continue initializing ProjectStore using the empty `initialState`. When `loadFromDisk` resolves, the initial call will find ProjectStore already initialized and the loaded state will never be used.

`accessProjects` uses `getProjectStore` concurrently for all projects so the risk of this bug is quite high. We will find a resolution for it in this PR before merging.

File: webapp/src/store/Store.ts
Comment (WULCAN) [inline webapp/src/store/Store.ts:46]: It is unsafe to use `writeFile` multiple times on the same file without waiting for the promise to be settled.

The probability of concurrent `updateTranslation` is high so we will have to find a resolution for this before merging.

File: webapp/src/store/Store.ts
Comment (WULCAN) [inline webapp/src/store/Store.ts:77]: `JSON.parse` could potentially execute for a long time if the input is very large and we never want to block Node's thread for long. Let's create a separate issue for exploring that and potentially finding an async replacement.

File: webapp/src/store/Store.ts
Comment (WULCAN) [inline webapp/src/store/Store.ts:80]: Hiding errors is going to make bugs hard to discover and diagnose. We should probably log this error so that instance maintainers can find it. We can defer that to a separate issue.

**Review by WULCAN — CHANGES_REQUESTED** [review]

File: webapp/src/store/Store.ts
Comment (WULCAN) [inline webapp/src/store/Store.ts:17]: Nope, keeping state in module variables does not work in Next. It seems server actions use a separate module tree, perhaps only in dev mode.

File: webapp/src/store/Store.ts
Comment (WULCAN) [inline webapp/src/store/Store.ts:61]: With this solution, we will never forget projects, even if they have been removed from `projects.yaml`

We will open a separate issue for that.

**Comment by WULCAN** [comment]
I have reviewed 13000e3, authored 6ac5ceb and I approve this for merging after someone has reviewed my changes between 13000e3 and 6ac5ceb

File: webapp/src/store/mergeStoreData.spec.ts
Comment (WULCAN) [inline webapp/src/store/mergeStoreData.spec.ts:145]: `toEqual` is not strict enough to catch this bug. It fins no difference between `{'any.message.id':undefined}` and `{}`.

Comment (amerharb) [reply webapp/src/store/mergeStoreData.spec.ts]: fixed by use toStrictEqual

**Comment by WULCAN** [comment]
Thank you, @amerharb 

I have reviewed up to and including 23ddabe1781fa17de482468476bba9fb96cedee0 and approve that for merging but I still want someone to review my changes 3a6dfc9ba2d919cd706d1ace03c18c0cbe030ab7 -- 6ac5ceb500fa24045ab42ab4a1930a0ccc7032bf

File: webapp/src/actions/updateTranslation.ts
Comment (amerharb) [inline webapp/src/actions/updateTranslation.ts]: This throw is not handled at caller

**Review by amerharb — APPROVED** [review]

**Review by WULCAN — APPROVED** [review]

File: webapp/src/actions/updateTranslation.ts
Comment (WULCAN) [reply webapp/src/actions/updateTranslation.ts]: I opened #146 for this.

**Pull request merged**
